hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java

```diff
@@ -17,103 +17,169 @@
+    if (GlobalTracer.isRegistered()) {
+      LOG.info("A tracer is already registered.");
+      return;
     }
 
-    if (tracer == null && conf != null) {
-      tracer = new Tracer.Builder("Tracer").conf(conf).build();
+    switch(c.get(HBASE_OPENTRACING_TRACER, HBASE_OPENTRACING_TRACER_DEFAULT)) {
+    case HBASE_OPENTRACING_TRACER_DEFAULT:
+      io.jaegertracing.Configuration conf = io.jaegertracing.Configuration.fromEnv(serviceName);
+      tracer = conf.getTracerBuilder().build();
+      break;
+    case HBASE_OPENTRACING_MOCKTRACER:
+      tracer = new MockTracer();
+      break;
+    default:
+      throw new RuntimeException("Unexpected tracer");
     }
+
+
+    GlobalTracer.register(tracer);
+
+  }
+
+  /*@VisibleForTesting
+  public static void registerTracerForTest(Tracer tracer) {
+    TraceUtil.tracer = tracer;
+    GlobalTracer.register(tracer);
+  }*/
+
+  public static Tracer getTracer() {
+    return tracer;
   }
 
   /**
-   * Wrapper method to create new TraceScope with the given description
-   * @return TraceScope or null when not tracing
+   * Wrapper method to create new Scope with the given description
+   * @return Scope or null when not tracing
    */
-  public static TraceScope createTrace(String description) {
-    return (tracer == null) ? null : tracer.newScope(description);
+  public static Scope createTrace(String description) {
+    return (tracer == null) ? null :
+        tracer.buildSpan(description).startActive(true);
   }
 
   /**
-   * Wrapper method to create new child TraceScope with the given description
+   * Wrapper method to create new child Scope with the given description
    * and parent scope's spanId
    * @param span parent span
-   * @return TraceScope or null when not tracing
+   * @return Scope or null when not tracing
    */
-  public static TraceScope createTrace(String description, Span span) {
-    if (span == null) {
-      return createTrace(description);
-    }
+  public static Scope createTrace(String description, Span span) {
+    if(span == null) return createTrace(description);
+
+    return (tracer == null) ? null : tracer.buildSpan(description).
+        asChildOf(span).startActive(true);
+  }
+
+  public static Scope createTrace(String description, SpanContext spanContext) {
+    if(spanContext == null) return createTrace(description);
 
-    return (tracer == null) ? null : tracer.newScope(description, span.getSpanId());
+    return (tracer == null) ? null : tracer.buildSpan(description).
+        asChildOf(spanContext).startActive(true);
   }
 
   /**
    * Wrapper method to add new sampler to the default tracer
    * @return true if added, false if it was already added
    */
-  public static boolean addSampler(Sampler sampler) {
+  public static boolean addSampler(SamplerConfiguration sampler) {
     if (sampler == null) {
       return false;
     }
 
-    return (tracer == null) ? false : tracer.addSampler(sampler);
+    conf = conf.withSampler(sampler);
```

Comment:
I think this is getting called from clients which needs the tracing. This 'conf' may be null I believe when this is called. @jojochuang - thanks for pointing to me to the docs. I was trying do some similar stuff and I took your patch as a reference. Now there are some quesitons - Assume I have kept my RegionServer to work with No sampling (just assume I don trace the flushes/compactions). Now if the client side I have enabled some sampling(tracing) say for eg writes -  if the writes are propogated to the server and we do the tracing - in that case will the sampling applied as what the client has decided ?

Reply:
@jojochuang - I have one more question. Say I  have already started my region server and in this case it is Jaegar tracing. But while I start I don want any real tracing to happen. For eg I need to have the NoopScopeManager. But now by default the Region Server will initiliaze with ThreadLocalScopeManager. 
`private ScopeManager scopeManager = new ThreadLocalScopeManager();`
The above is from JaegerTracer.Builder.
Now if I really want a tracer but with the scope doing nothing - is it possible? Next say if I want to dynamically enable tracing and then disable it how is that possible. 
The config file that you had pointed out to is mainly to define per service what is the sampling type I want. But that again is predefined and there is no changing it. My requirement is if the RS needs to run with minimum friction and only on demand do my trace and again go back to minimum friction is it possible?